include/xtensor/xpad.hpp

```diff
@@ -81,6 +81,25 @@ namespace xt
 
         using size_type = typename std::decay_t<E>::size_type;
         using value_type = typename std::decay_t<E>::value_type;
+        using type = typename std::remove_reference<E&>::type;
```

Comment:
`using return_type = std::decay_t<E>` will fix the compilation issue with constant expressions. However this code would work with assignable expressions only (not unevaluated `xfunction` for instance), which is the current behavior.

OR

```cpp
using return_type = temporary_type_t<value_type,
                                     typename std::decay_t<E>::shape_type,
                                     std::decay_t<E>::static_layout>;
```
will fix the issue for both constant assignlable expressions and unevaluated expressions.

In both case, I find `return_type` more meaningful than `type`.